hudi-client/src/test/java/org/apache/hudi/execution/bulkinsert/TestBulkInsertInternalPartitioner.java

```diff
@@ -0,0 +1,143 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.apache.hudi.common.testutils.HoodieTestDataGenerator.newHoodieRecords;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class TestBulkInsertInternalPartitioner extends HoodieClientTestBase {
+
+  public static JavaRDD<HoodieRecord> generateTestRecordsForBulkInsert(JavaSparkContext jsc)
+      throws Exception {
+    // RDD partition 1
+    List<HoodieRecord> records1 = newHoodieRecords(3, "2020-07-31T03:16:41.415Z");
+    records1.addAll(newHoodieRecords(2, "2020-08-01T03:16:41.415Z"));
+    records1.addAll(newHoodieRecords(5, "2020-07-31T03:16:41.415Z"));
+    // RDD partition 2
+    List<HoodieRecord> records2 = newHoodieRecords(4, "2020-08-02T03:16:22.415Z");
+    records2.addAll(newHoodieRecords(1, "2020-07-31T03:16:41.415Z"));
+    records2.addAll(newHoodieRecords(5, "2020-08-01T06:16:41.415Z"));
+    return jsc.parallelize(records1, 1).union(jsc.parallelize(records2, 1));
+  }
+
+  public static Map<String, Long> generateExpectedPartitionNumRecords() {
+    Map<String, Long> expectedPartitionNumRecords = new HashMap<>();
+    expectedPartitionNumRecords.put("2020/07/31", 9L);
+    expectedPartitionNumRecords.put("2020/08/01", 7L);
+    expectedPartitionNumRecords.put("2020/08/02", 4L);
+    return expectedPartitionNumRecords;
+  }
+
+  private static JavaRDD<HoodieRecord> generateTripleTestRecordsForBulkInsert(JavaSparkContext jsc)
+      throws Exception {
+    return generateTestRecordsForBulkInsert(jsc).union(generateTestRecordsForBulkInsert(jsc))
+        .union(generateTestRecordsForBulkInsert(jsc));
+  }
+
+  public static Map<String, Long> generateExpectedPartitionNumRecordsTriple() {
+    Map<String, Long> expectedPartitionNumRecords = generateExpectedPartitionNumRecords();
+    for (String partitionPath : expectedPartitionNumRecords.keySet()) {
+      expectedPartitionNumRecords.put(partitionPath,
+          expectedPartitionNumRecords.get(partitionPath) * 3);
+    }
+    return expectedPartitionNumRecords;
+  }
+
+  private static Stream<Arguments> configParams() {
+    Object[][] data = new Object[][] {
+        {BulkInsertInternalPartitioner.BulkInsertSortMode.GLOBAL_SORT, true, true},
+        {BulkInsertInternalPartitioner.BulkInsertSortMode.PARTITION_SORT, false, true},
+        {BulkInsertInternalPartitioner.BulkInsertSortMode.NONE, false, false}
+    };
+    return Stream.of(data).map(Arguments::of);
+  }
+
+  private void verifyRecordAscendingOrder(Iterator<HoodieRecord> records) {
+    HoodieRecord prevRecord = null;
+
+    for (Iterator<HoodieRecord> it = records; it.hasNext(); ) {
```

Comment:
would n't it be simpler to compare two sorted lists for equivalency? instead of hand-crafting this logic of comparing for ascending order?

Reply:
Yes, make sense.  Fixed.